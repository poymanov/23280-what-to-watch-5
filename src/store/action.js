export const ActionType = {
  LOAD_MOVIES: `LOAD_MOVIES`,
  LOAD_GENRES: `LOAD_GENRES`,
  LOAD_PROMO_MOVIE: `LOAD_PROMO_MOVIE`,
  LOAD_USER: `LOAD_USER`,
  LOAD_AUTH_FORM_ERROR: `LOAD_AUTH_FORM_ERROR`,
  LOAD_MOVIES_PAGINATION: `LOAD_MOVIES_PAGINATION`,
  LOAD_CURRENT_MOVIE: `LOAD_CURRENT_MOVIE`,
  LOAD_MOVIE_REVIEWS: `LOAD_MOVIE_REVIEWS`,
  LOAD_USER_FAVORITES: `LOAD_USER_FAVORITES`,
  LOAD_MOVIE_RELATED: `LOAD_MOVIE_RELATED`,
  FLUSH_CURRENT_MOVIE: `FLUSH_CURRENT_MOVIE`,
  CHANGE_GENRE_FILTER: `CHANGE_GENRE_FILTER`,
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  REDIRECT_TO_ROUTE: `REDIRECT_TO_ROUTE`,
};

export const loadMovies = (movies) => ({
  type: ActionType.LOAD_MOVIES,
  payload: movies,
});

export const loadGenres = (movies) => ({
  type: ActionType.LOAD_GENRES,
  payload: movies,
});

export const loadPromoMovie = (movie) => ({
  type: ActionType.LOAD_PROMO_MOVIE,
  payload: movie,
});

export const loadCurrentMovie = (movie) => ({
  type: ActionType.LOAD_CURRENT_MOVIE,
  payload: movie,
});

export const loadMovieReviews = (reviews) => ({
  type: ActionType.LOAD_MOVIE_REVIEWS,
  payload: reviews,
});

export const loadMovieRelated = (movies) => ({
  type: ActionType.LOAD_MOVIE_RELATED,
  payload: movies,
});

export const loadUserFavorites = (movies) => ({
  type: ActionType.LOAD_USER_FAVORITES,
  payload: movies,
});

export const flushCurrentMovie = () => ({
  type: ActionType.FLUSH_CURRENT_MOVIE,
});

export const changeGenreFilter = (genreId) => ({
  type: ActionType.CHANGE_GENRE_FILTER,
  payload: genreId,
});

export const requireAuthorization = (status) => ({
  type: ActionType.REQUIRED_AUTHORIZATION,
  payload: status,
});

export const loadUser = (user) => ({
  type: ActionType.LOAD_USER,
  payload: user,
});

export const loadAuthFormError = (error) => ({
  type: ActionType.LOAD_AUTH_FORM_ERROR,
  payload: error,
});

export const redirectToRoute = (url) => ({
  type: ActionType.REDIRECT_TO_ROUTE,
  payload: url,
});
