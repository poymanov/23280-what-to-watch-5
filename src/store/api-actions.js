import {
  loadGenres, loadMovies, loadPromoMovie, loadCurrentMovie,
  requireAuthorization, redirectToRoute,
  loadUser, loadMovieReviews, loadUserFavorites, loadMovieRelated, loadAuthFormError
} from "./action";
import {AuthorizationStatus, AppRoute, APIRoute} from "../const";

export const fetchMoviesList = () => (dispatch, _getState, api) => (
  api.get(APIRoute.FILMS)
    .then(({data}) => {
      dispatch(loadMovies(data));
      return {data};
    })
    .then(({data}) => {
      dispatch(loadGenres(data));
      return {data};
    })
);

export const fetchPromoMovie = () => (dispatch, _getState, api) => (
  api.get(APIRoute.FILMS_PROMO)
    .then(({data}) => dispatch(loadPromoMovie(data)))
);

export const fetchCurrentMovie = (movieId) => (dispatch, _getState, api) => (
  api.get(APIRoute.FILMS + `/${movieId}`)
    .then(({data}) => dispatch(loadCurrentMovie(data)))
    .then(() => {
      api.get(APIRoute.FILMS)
        .then(({data}) => {
          dispatch(loadMovieRelated(data));
        });
    })
);

export const fetchMovieReviews = (movieId) => (dispatch, _getState, api) => (
  api.get(APIRoute.MOVIE_REVIEWS + `/${movieId}`)
    .then(({data}) => dispatch(loadMovieReviews(data)))
);

export const fetchUserFavorites = () => (dispatch, _getState, api) => (
  api.get(APIRoute.FAVORITE)
    .then(({data}) => dispatch(loadUserFavorites(data)))
);

export const checkAuth = () => (dispatch, _getState, api) => (
  api.get(APIRoute.LOGIN)
    .then(({data}) => {
      dispatch(requireAuthorization(AuthorizationStatus.AUTH));

      return {data};
    })
    .then(({data}) => dispatch(loadUser(data)))
    .catch(() => {})
);

export const login = ({login: email, password}) => (dispatch, _getState, api) => (
  api.post(APIRoute.LOGIN, {email, password})
    .then(({data}) => {
      dispatch(requireAuthorization(AuthorizationStatus.AUTH));
      return {data};
    })
    .then(({data}) => dispatch(loadUser(data)))
    .then(() => dispatch(redirectToRoute(AppRoute.ROOT)))
    .catch(({response}) => dispatch(loadAuthFormError(response.data.error)))
);

export const addReview = ({id, rating, comment}) => (dispatch, _getState, api) => (
  api.post(APIRoute.MOVIE_REVIEWS + `/${id}`, {rating, comment})
    .then(() => dispatch(redirectToRoute(AppRoute.FILMS + `/${id}`)))
);

export const addMovieToFavorite = (id) => (dispatch, _getState, api) => (
  api.post(APIRoute.FAVORITE + `/${id}/1`)
    .then(() => dispatch(redirectToRoute(AppRoute.FILMS + `/${id}`)))
);
