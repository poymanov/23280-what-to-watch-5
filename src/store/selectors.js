import {NameSpace} from "./reducers/root-reducer";

export const mainMoviesSelector = (state) => {
  return state[NameSpace.MOVIES].main;
};

export const myListMoviesSelector = (state) => {
  return state[NameSpace.MOVIES].main;
};

export const relatedMoviesSelector = (state) => {
  return state[NameSpace.MOVIES].main;
};

export const promoMovieSelector = (state) => {
  return state[NameSpace.MOVIES].promo;
};

export const currentFilterIdSelector = (state) => {
  return state[NameSpace.MOVIES].filterGenreId;
};

