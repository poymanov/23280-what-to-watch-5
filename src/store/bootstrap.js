import {createAPI} from "../services/api";
import {requireAuthorization} from "./action";
import {AuthorizationStatus} from "../constants/const";
import {applyMiddleware, createStore} from "redux";
import rootReducer from "../store/reducers/root-reducer";
import {composeWithDevTools} from "redux-devtools-extension";
import thunk from "redux-thunk";
import {redirect} from "./middlewares/redirect";

export const initStore = () => {
  const api = createAPI(
      () => store.dispatch(requireAuthorization(AuthorizationStatus.NO_AUTH))
  );

  const store = createStore(
      rootReducer,
      composeWithDevTools(
          applyMiddleware(thunk.withExtraArgument(api)),
          applyMiddleware(redirect)
      )
  );

  return store;
};
