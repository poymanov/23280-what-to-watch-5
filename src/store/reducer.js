import {extend} from "../utils";
import {ActionType} from "./action";
import genres from "../mocks/genres";
import movies from "../mocks/movies";
import {filterMoviesByGenreId, paginateMovies} from "../movies";

const initialState = {
  filterGenreId: 0,
  genres,
  moviesByGenre: paginateMovies(filterMoviesByGenreId(movies, 0), 0)
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_MOVIES_FILTER:
      const payload = action.payload;
      return extend(state, {
        filterGenreId: payload.genreId,
        moviesByGenre: payload.filteredMovies
      });
    case ActionType.SHOW_MORE_MOVIES:
      return extend(state, {
        moviesByGenre: action.payload.nextMovies
      });
  }

  return state;
};

export {reducer};
