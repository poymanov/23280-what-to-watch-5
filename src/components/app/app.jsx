import React from "react";
import {Route, Switch, Router as BrowserRouter} from "react-router-dom";
import Main from "../main/main";
import SignIn from "../sign-in/sign-in";
import MyList from "../my-list/my-list";
import Movie from "../movie/movie";
import AddReview from "../add-review/add-review";
import Player from "../player/player";
import withFullscreen from "../../hocs/with-fullscreen";
import withPlayerControls from "../../hocs/with-player-controls";
import {connect} from "react-redux";
import {promoMovieSelector} from "../../store/selectors";
import PrivateRoute from "../private-route/private-route";
import MovieTypes from "../../types/movies";
import browserHistory from "../../browser-history";

const App = (props) => {
  const {movie} = props;

  return (
    <BrowserRouter history={browserHistory}>
      <Switch>
        <Route path="/" exact render={({history}) => (
          <Main
            onPlayButtonClick={(movieId) => history.push(`/player/` + movieId)}
          />
        )} />
        <Route path="/login" exact>
          <SignIn />
        </Route>
        <Route path="/my-list" exact>
          <MyList />
        </Route>
        <Route path="/films/:id" exact render={({history, match}) => (
          <Movie id={match.params.id} onPlayButtonClick={(movieId) => history.push(`/player/` + movieId)} />
        )} />
        <PrivateRoute path="/films/:id/review" exact render={() => <AddReview />} />
        <Route path="/player/:id" exact render={({history}) => {
          const PlayerWrapped = withFullscreen(withPlayerControls(Player));

          return <PlayerWrapped
            movie={movie}
            onPlayerClose={() => history.goBack()}
          />;
        }} />
      </Switch>
    </BrowserRouter>
  );
};

App.propTypes = {
  movie: MovieTypes.item,
};

const mapStateToProps = (state) => ({
  movie: promoMovieSelector(state)
});

export {App};
export default connect(mapStateToProps)(App);
