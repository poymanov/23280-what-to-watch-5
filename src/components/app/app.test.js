import React from "react";
import renderer from "react-test-renderer";
import {Router as BrowserRouter} from "react-router-dom";
import {Provider} from "react-redux";
import {initStore} from "../../store/bootstrap";
import browserHistory from "../../etc/browser-history";
import {App} from "./app";

const movie = {
  id: 1,
  name: `test`,
  posterImage: `test`,
  previewImage: `test`,
  backgroundImage: `test`,
  backgroundColor: `test`,
  videoLink: `test`,
  previewVideoLink: `test`,
  description: `test`,
  rating: 1,
  scoresCount: 1,
  director: `test`,
  starring: [`test`],
  runTime: 1,
  genre: `test`,
  released: 1,
  isFavorite: false,
};

it(`Should App render correctly`, () => {
  const tree = renderer
    .create(
        <Provider store={initStore()}>
          <BrowserRouter history={browserHistory}>
            <App movie={movie} />
          </BrowserRouter>
        </Provider>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
