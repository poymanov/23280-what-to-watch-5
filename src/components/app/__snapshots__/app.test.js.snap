// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should App render correctly 1`] = `
Array [
  <section
    className="movie-card"
  >
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header user-page__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="user-block"
      >
        <a
          className="user-block__link"
          href="/login"
          onClick={[Function]}
        >
          Sign in
        </a>
      </div>
    </header>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      />
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
